File: open3d_manage/Method/normal.py
import numpy as np
import open3d as o3d
from typing import Union

def toNormals(points: Union[np.ndarray, list], knn_num: int = 0, need_smooth: bool = False) -> np.ndarray:
    if isinstance(points, list):
        points = np.array(points)

    is_flatten = len(points.shape) == 1

    if is_flatten:
        points = points.reshape(-1, 3)

    pcd = o3d.geometry.PointCloud()
    pcd.points = o3d.utility.Vector3dVector(points)

    if knn_num > 0:
        pcd.estimate_normals(o3d.geometry.KDTreeSearchParamKNN(knn_num))
    else:
        pcd.estimate_normals()

    pcd.normalize_normals()

    if need_smooth > 0:
        pcd.orient_normals_consistent_tangent_plane(knn_num)

    normals = np.asarray(pcd.normals)

    if is_flatten:
        normals = normals.reshape(-1)

    return normals

def toNormalsList(points: list, knn_num: int = 0, need_smooth: bool = False) -> list:
    return toNormals(points, knn_num, need_smooth).tolist()

File: src/normal.cpp
#include "normal.h"
#include "pcd.h"
#include <Eigen/Core>
#include <open3d/Open3D.h>

const nc::NdArray<float> toNormals(const nc::NdArray<float> &points,
                                   const int &knn_num,
                                   const bool &need_smooth) {
  const int point_num = points.shape().rows;

  open3d::geometry::PointCloud pcd = toPcd(points);

  if (knn_num > 0) {
    pcd.EstimateNormals(open3d::geometry::KDTreeSearchParamKNN(knn_num));
  } else {
    pcd.EstimateNormals();
  }

  pcd.NormalizeNormals();

  if (need_smooth > 0) {
    pcd.OrientNormalsConsistentTangentPlane(knn_num);
  }

  nc::NdArray<float> normals = nc::zeros<float>(point_num, 3);

  for (int i = 0; i < point_num; ++i) {
    normals(i, 0) = pcd.normals_[i][0];
    normals(i, 1) = pcd.normals_[i][1];
    normals(i, 2) = pcd.normals_[i][2];
  }

  return normals;
}
